### PR TITLE
fix(modal): compile modal before inserting into DOM

### LINKED ISSUE #5183: display issues when opening modal
reproducible on demo: http://angular-ui.github.io/bootstrap/#/modal

issue 1:
open modal and pay attention to the page in the background. notice it shifts down briefly before returning to normal.

issue 2:
turn off animation. open a large or small modal and pay attention to how the modal content area displays as its opened. notice it shrinks or expands before displaying as expected.

### PR DESCRIPTION
**Pull request by onumossn**
I think this fixes #5183

### CODE REVIEW

**Comment by wesleycho** [comment]
This looks like a good fix to me - thoughts @Foxandxss and/or @icfantv ?

**Comment by Foxandxss** [comment]
If it fixes it, I am all in, but we need to be sure (AKA plunker).

**Comment by icfantv** [comment]
Looks like there's 13 broken tests in the build.  Need to see if the tests are no longer valid or they need to be updated.  @onumossn, if you would like this merged, your work is not yet complete.

**Comment by onumossn** [comment]
The original commit, actually, only fixed issue 1 of the two listed on #5183, so I made additional updates to fix the 2nd issue.

@Foxandxss The cases can be tested from the demos, so is a plunker necessary?

@icfantv Sorry about that. I had used the wrong command to test initially. I updated the tests. Previously an extra $animate.flush was used because the $compile would use more $animates after the first $animate.flush in the open function, but since the compile no longer waits for the first animation, its no longer necessary.

**Comment by Foxandxss** [comment]
A test to prove that your PR fixes the issue.

**Comment by icfantv** [comment]
@onumossn, what @Foxandxss means is take the proper file from the `dist` folder and create a plunker using it.  Easiest thing to do would be to use the modal plunker demo (from the docs page) remove the script include for UIBS, and add yours locally instead.

@Foxandxss, @wesleycho, IIRC we added those `flush()` calls when angular-animate changed between 1.4.4 and 1.4.5, right?

**Comment by wesleycho** [comment]
The `flush()` has to do with ngAnimate.

What you could do is test for the newly created element's presence in the element being appended to (`body` by default) and then flush.

**Comment by onumossn** [comment]
`flush` is being run in the `open` helper. However, `flush` needed to be run twice before because after the first `flush` the `uib-modal-window` directive gets compiled which adds more animation to the buffer which needs to be flushed again. In this version, the compile occurs before the first `flush` so all the animations are in the buffer when the first `flush` occurs in the `open` helper. I may not have been very clear in my last comment.

before plunker:
http://embed.plnkr.co/YbWBhcwoFECDFnpP4rnW/

after plunker:
http://embed.plnkr.co/pCBkbYY1BIkBeQ8lUR0q/

**Comment by onumossn** [comment]
before plunker: 
https://embed.plnkr.co/YbWBhcwoFECDFnpP4rnW/

after plunker:
https://embed.plnkr.co/pCBkbYY1BIkBeQ8lUR0q/

File: template/modal/window.html [hunk 1]
```diff
@@ -2,5 +2,5 @@
     uib-modal-animation-class="fade"
     modal-in-class="in"
     ng-style="{'z-index': 1050 + index*10, display: 'block'}">
-    <div class="modal-dialog" ng-class="size ? 'modal-' + size : ''"><div class="modal-content" uib-modal-transclude></div></div>
+    <div class="modal-dialog {{size ? 'modal-' + size : ''}}"><div class="modal-content" uib-modal-transclude></div></div>
```
Comment (wesleycho) [inline template/modal/window.html]: Why is this change made? This seems unnecessary

**Comment by wesleycho** [comment]
This needs a test still that the element is compiled before the $animate.enter is fired.

File: template/modal/window.html (hunk 1 above)
Comment (onumossn) [inline template/modal/window.html]: ngClass was going through the animation processes of all the before and after classes which was causing the size when picking small or large to be regular size when initially rendered and the jump to the correct size.

Comment (wesleycho) [inline template/modal/window.html]: Ugh, that's a negative of it being tied to $animate...but I guess it can't be avoided then.

**Comment by onumossn** [comment]
Added tests

File: src/modal/test/modal.spec.js [hunk 2]
```diff
@@ -261,6 +265,46 @@ describe('$uibModal', function () {
...
+        scope: $rootScope.$new(),
+        deferred: modalInstance.result,
+        renderDeferred: modalInstance.rendered,
+        closedDeferred: modalInstance.closed,
+        content: '<div id="test">{{\'' + expectedText + '\'}}</div>'
+      });
+
+      topModal = $uibModalStack.getTop();
+
+      expect(topModal.value.modalDomEl[0].querySelector('#test')).toBeNull();
```
Comment (wesleycho) [inline src/modal/test/modal.spec.js]: You could just do `topModal.value.modalDomEl.find('#test')` here I believe - we have jQuery in the testing situation.

File: src/modal/test/modal.spec.js [hunk 3]
```diff
@@ -261,6 +265,46 @@ describe('$uibModal', function () {
...
+      expect(document.querySelector('#test')).toBeNull();
```
Comment (wesleycho) [inline src/modal/test/modal.spec.js]: You can just do `$('#test').length`

File: src/modal/test/modal.spec.js [hunk 4]
```diff
@@ -261,6 +265,46 @@ describe('$uibModal', function () {
...
+
+      $rootScope.$digest();
+
+      expect(topModal.value.modalDomEl[0].querySelector('#test').textContent).toEqual(expectedText);
```
Comment (wesleycho) [inline src/modal/test/modal.spec.js]: Ditto on the query selector

File: src/modal/test/modal.spec.js [hunk 5]
```diff
@@ -261,6 +265,46 @@ describe('$uibModal', function () {
...
+      expect(document.querySelector('#test').textContent).toEqual(expectedText);
```
Comment (wesleycho) [inline src/modal/test/modal.spec.js]: Could just do `$('#test').text()`

**Comment by wesleycho** [comment]
Once the query selectors are changed to just use jQuery, can you squash your commits into one?

This LGTM otherwise.

**Comment by onumossn** [comment]
Update the tests to use jQuery. Sorry for the delay.

**Comment by wesleycho** [comment]
LGTM - thanks for this!

**Pull request closed without merging**